File: 0567 Permutation in String.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if not s1 or not s2 or len(s1) > len(s2): return False
        s1h = {}
        for c in 'abcdefghijklmnopqrstuvwxyz':
            s1h[c] = 0
        for c in s1:
            s1h[c] += 1
        cnt, l, r = len(s1), 0, 0
        while r < len(s2):
            if s1h[s2[r]] > 0: cnt -= 1
            s1h[s2[r]] -= 1
            r += 1
            if cnt == 0: return True
            if r - l == len(s1):
                if s1h[s2[l]] >= 0: cnt += 1
                s1h[s2[l]] += 1
                l += 1
        return False

File: 0567 Permutation in String.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        if (s1 == null || s2 == null || s1.length() > s2.length()) return false;
        int[] s1h = new int[26];
        for (char c : s1.toCharArray()) ++s1h[c - 'a'];
        int l = 0, r = 0, cnt = s1.length();
        while (r < s2.length()) {
            if (s1h[s2.charAt(r++) - 'a']-- > 0) --cnt;
            if (cnt == 0) return true;
            if (r - l == s1.length() && s1h[s2.charAt(l++) - 'a']++ >= 0) ++cnt;
        }
        return false;
    }
}